File: A3/class_picture.cpp
/*https://ualberta.kattis.com/sessions/zy3te5/problems/classpicture
*/


#include <bits/stdc++.h>

using namespace std;

bool works(vector<int>& permutation, vector<vector<bool>>& conflicts) {
    for(int i = 1; i < permutation.size(); i++)
        if(conflicts[permutation[i-1]][permutation[i]])
            return false;
    return true;
}

void class_picture(vector<string>& names, vector<vector<bool>>& conflicts) {
  vector<int> permutation;
  for(int i = 0; i < names.size(); i++) {
      permutation.push_back(i);
  }

  do {
      if(works(permutation, conflicts)) {
        for(int i : permutation) 
          cout << names[i] << " ";
        cout << endl;
        return;
      }
  } while(next_permutation(permutation.begin(), permutation.end()));

  cout << "You all need therapy." << endl;
}

int main() {
  int n;
  while (cin >> n) {
    vector<string> names;
    for(int i = 0; i < n; i++) {
      string s;
      cin >> s;
      names.push_back(s);
    }
    sort(names.begin(), names.end());

    vector<vector<bool>> conflicts;
    conflicts.resize(n, vector<bool>(n, false));

    int m;
    cin >> m;
    while(m--) {
      string s1, s2;
      cin >> s1 >> s2;
      int index1 = find(names.begin(), names.end(), s1) - names.begin();
      int index2 = find(names.begin(), names.end(), s2) - names.begin();
      conflicts[index1][index2] = true;
      conflicts[index2][index1] = true;
    }
    class_picture(names, conflicts);
  }
}

File: brute_force/class_picture.py
"""https://ualberta.kattis.com/sessions/zy3te5/problems/classpicture
"""

import itertools
import sys
from typing import Dict, List, Set, Tuple

def construct_enemies_map(enemy_pairs: List[Tuple[str]]) -> Dict[str, Set[str]]:
    enemies_map = {}
    for enemy_pair in enemy_pairs:
        enemy1, enemy2 = enemy_pair
        if enemy1 not in enemies_map:
            enemies_map[enemy1] = set()
        enemies_map[enemy1].add(enemy2)
        if enemy2 not in enemies_map:
            enemies_map[enemy2] = set()
        enemies_map[enemy2].add(enemy1)
    return enemies_map

def class_picture(class_students: List[str], enemy_pairs: List[Tuple[str]]) -> None:
    enemies_map = construct_enemies_map(enemy_pairs)
    possible_orderings = list(itertools.permutations(sorted(class_students)))
    for possible_ordering in possible_orderings:
        found = True
        for i in range(1, len(possible_ordering)):
            person1 = possible_ordering[i-1]
            person2 = possible_ordering[i]
            if person1 in enemies_map.get(person2, set()):
                found = False
                break
        if found:
            print(' '.join(possible_ordering))
            return
    print('You all need therapy.')

def main():
    input_lines = sys.stdin.read().splitlines()
    i = 0
    while i < len(input_lines):
        class_students = []
        num_students = int(input_lines[i])
        i += 1
        start = i
        while i < num_students + start:
            class_students.append(input_lines[i])
            i += 1
        num_of_enemy_pairs = int(input_lines[i])
        i += 1
        enemy_pairs = []
        start = i
        while i < num_of_enemy_pairs + start:
            enemy_pairs.append(tuple(input_lines[i].split()))
            i += 1
        class_picture(class_students, enemy_pairs)

    # class_picture(['Ron', 'George', 'Bill', 'Fred', 'Jenny'],
    #     [('Fred', 'Jenny'), ('Bill', 'Ron'), ('George', 'Jenny')]
    # )
    # class_picture(['Alice', 'Bob'], [('Alice', 'Bob')])

if __name__ == '__main__':
    main()